src/core/jbig2.js

```diff
@@ -607,6 +607,136 @@ var Jbig2Image = (function Jbig2ImageClosure() {
+        for (x = xMin, subX = 0; x < xMax; x++, subX++) {
+          row[subX] = collectiveBitmap[y][x];
+        }
+      }
+      patterns[i] = patternBitmap;
+      i++;
+    }
+    return patterns;
+  }
+
+  function decodeHalftoneRegion(mmr, patterns, template, regionWidth,
+                                regionHeight, defaultPixelValue, enableSkip,
+                                combinationOperator, gridWidth, gridHeight,
+                                gridOffsetX, gridOffsetY, gridVectorX,
+                                gridVectorY, decodingContext) {
+    let skip = null;
+    if (enableSkip) {
+      error('JBIG2 error: skip is not supported');
+    }
+    if (combinationOperator !== 0) {
+      error('JBIG2 error: operator ' + combinationOperator +
+        ' is not supported in halftone region');
+    }
+
+    // Prepare bitmap.
+    let regionBitmap = [];
+    let i, j, row;
+    for (i = 0; i < regionHeight; i++) {
+      row = new Uint8Array(regionWidth);
+      if (defaultPixelValue) {
+        for (j = 0; j < regionWidth; j++) {
+          row[j] = defaultPixelValue;
+        }
+      }
+      regionBitmap.push(row);
+    }
+
+    let numberOfPatterns = patterns.length;
+    let pattern0 = patterns[0];
+    let patternWidth = pattern0[0].length, patternHeight = pattern0.length;
+    let bitsPerValue = log2(numberOfPatterns);
+    let at = [];
+    at.push({
+      x: (template <= 1) ? 3 : 2,
+      y: -1,
+    });
+    if (template === 0) {
+      at.push({
+        x: -3,
+        y: -1,
+      });
+      at.push({
+        x: 2,
+        y: -2,
+      });
+      at.push({
+        x: -2,
+        y: -2,
+      });
+    }
+    // Annex C. Gray-scale Image Decoding Procedure.
+    let grayScaleBitPlanes = [];
+    for (i = bitsPerValue - 1; i >= 0; i--) {
+      grayScaleBitPlanes[i] = decodeBitmap(mmr, gridWidth, gridHeight,
+                                           template, false, skip, at,
+                                           decodingContext);
+    }
+    // 6.6.5.2 Rendering the patterns.
+    let mg, ng, bit, patternIndex, patternBitmap, x, y, patternRow, regionRow;
+    for (mg = 0; mg < gridHeight; mg++) {
+      for (ng = 0; ng < gridWidth; ng++) {
+        bit = 0;
+        patternIndex = 0;
+        for (j = bitsPerValue - 1; j >= 0; j--) {
+          bit = grayScaleBitPlanes[j][mg][ng] ^ bit; // Gray decoding
+          patternIndex |= bit << j;
+        }
+        patternBitmap = patterns[patternIndex];
+        x = (gridOffsetX + mg * gridVectorY + ng * gridVectorX) >> 8;
+        y = (gridOffsetY + mg * gridVectorX - ng * gridVectorY) >> 8;
+        // Draw patternBitmap at (x, y).
+        for (i = 0; i < patternHeight; i++) {
+          regionRow = regionBitmap[y + i];
+          if (!regionRow) {
+            continue;
+          }
+          patternRow = patternBitmap[i];
+          for (j = 0; j < patternWidth; j++) {
+            regionRow[x + j] |= patternRow[j]; // OR
+          }
+        }
```

Comment:
In the past we had performance issues when array is accessed out-of-bounds. Can you create if-check for fast version of copy:

```
  if (x >= 0 && x + patternWidth <= regionWidth && y >= 0 && y + patternHeight <= regionHeight) {
     for (i = 0; i < patternHeight; i++) {
        regionRow = regionBitmap[y + i];
        for (j = 0; j < patternWidth; j++) {
            regionRow[x + j] |= patternRow[j]; // OR
        }
      }
  } else {
     // same loop as above but y + i and x + j is checked for out-of-bound or for-loop use proper bounds
  }
```